### PR TITLE
feat: harden write buffer aggregator (#3805)

### PR DESCRIPTION
**Pull request by tustvold**
Manual inspection doesn't seem to be yielding insight into what is going on, so lets add some more assertions and logs and hope they shed some more light on the situation.

### CODE REVIEW

File: write_buffer/src/kafka/aggregator.rs [hunk 1]
```diff
@@ -464,6 +500,15 @@ impl StatusDeaggregator for Deaggregator {
...
     ) -> Result<Self::Status, aggregator::Error> {
+        assert_eq!(input.len(), self.tag_to_record.len(), "invalid offsets");
+        assert!(
+            self.tag_to_record.len() > tag.0,
+            "tag {} out of range (tag_to_record: {:?}, offsets: {:?})",
+            tag.0,
+            self.tag_to_record,
+            input
+        );
+
```
Comment (tustvold) [inline write_buffer/src/kafka/aggregator.rs:511]: The line below is the one that is panic-ing. At least now we should at the very least gain the context of what the `Deaggregator` state is

File: write_buffer/src/kafka/aggregator.rs [hunk 2]
```diff
@@ -370,9 +375,40 @@ impl Aggregator for DmlAggregator {
...
         state.flush_writes();
 
+        assert_eq!(self.state.current_writes.len(), 0, "incomplete flush");
+        assert_eq!(
+            self.state.completed_ops.len(),
+            self.state.tag_to_record.len(),
+            "missing records following flush"
+        );
+
+        info!(
```
Comment (tustvold) [inline write_buffer/src/kafka/aggregator.rs:405]: This log line may be rather spammy, but at this point the more context the better - we can always lower its level later

File: write_buffer/src/kafka/aggregator.rs [hunk 3]
```diff
@@ -322,20 +329,18 @@ impl Aggregator for DmlAggregator {
...
-                            let mut md2 = op_md;
-                            std::mem::swap(agg, &mut agg2);
-                            std::mem::swap(record, &mut record2);
-                            std::mem::swap(md, &mut md2);
 
-                            let flushed_tag = agg2.tag;
-                            agg2.finalize_span();
-                            self.state.push_op(record2, md2, flushed_tag);
+                            // Replace current aggregator
+                            let (agg, record, md) = o.insert((new_agg, op_record, op_md));
```
Comment (tustvold) [inline write_buffer/src/kafka/aggregator.rs:340]: Just a drive by refactor to make it clearer

**Review by crepererum — APPROVED** [review]
Looks good. You might wanna put a bit of logging into `reserve_tag` as well. This is (ensured by the `Tag` type) the only place where tags are created, which is exactly the tag that later on in `deaggregate` `self.tag_to_record[tag.0]` panics.

**Comment by tustvold** [comment]
> You might wanna put a bit of logging into reserve_tag as well.

I think this should already be covered by the existing length assertions, as the tags are just based on the length of `tag_to_record` but I could be mistaken?? I'll run with this for now, and see what happens

**Pull request merged**